server/shadow/shadow_client.c

```diff
@@ -385,6 +378,64 @@ BOOL shadow_client_activate(freerdp_peer* peer)
 	return shadow_client_refresh_rect(client, 0, NULL);
 }
 
+BOOL shadow_client_logon(freerdp_peer* peer, SEC_WINNT_AUTH_IDENTITY* identity, BOOL automatic)
+{
+	char* user = NULL;
+	char* domain = NULL;
+	char* password = NULL;
+	rdpSettings* settings = peer->settings;
+
+	if (identity->Flags & SEC_WINNT_AUTH_IDENTITY_UNICODE)
+	{
+		if (identity->User)
+			ConvertFromUnicode(CP_UTF8, 0, identity->User, identity->UserLength, &user, 0, NULL, NULL);
+
+		if (identity->Domain)
+			ConvertFromUnicode(CP_UTF8, 0, identity->Domain, identity->DomainLength, &domain, 0, NULL, NULL);
+
+		if (identity->Password)
+			ConvertFromUnicode(CP_UTF8, 0, identity->Password, identity->PasswordLength, &user, 0, NULL, NULL);
+	}
+	else
+	{
+		if (identity->User)
+			user = _strdup((char*) identity->User);
+
+		if (identity->Domain)
+			domain = _strdup((char*) identity->Domain);
+
+		if (identity->Password)
+			password = _strdup((char*) identity->Password);
+	}
+
+	if (user)
+	{
+		free(settings->Username);
+		settings->Username = user;
+		user = NULL;
+	}
+
+	if (domain)
+	{
+		free(settings->Domain);
+		settings->Domain = domain;
+		user = NULL;
+	}
+
+	if (password)
+	{
+		free(settings->Password);
+		settings->Password = password;
+		password = NULL;
+	}
+
+	free(user);
```

Comment:
I think these frees are not necessary as none of the 3 vars will be non-NULL at this point